File: ts/test/alphabetCipher.test.ts
import { describe, it, expect } from "vitest";
import { encode, decode } from "../src/alphabetCipher";

describe("Alphabet Cipher", () => {
  it("encodes the sample from the prompt", () => {
    expect(encode("meetmebythetree", "scones")).toBe("egsgqwtahuiljgs");
  });

  it("decodes the sample from the prompt", () => {
    expect(decode("egsgqwtahuiljgs", "scones")).toBe("meetmebythetree");
  });

  it("round-trips for simple inputs", () => {
    const msg = "helloworld";
    const key = "abc";
    const c = encode(msg, key);
    expect(decode(c, key)).toBe(msg);
  });

  it("normalizes case (lowercases both message and keyword)", () => {
    expect(encode("MeetMeByTheTree", "SCONES")).toBe("egsgqwtahuiljgs");
  });

  it("ignores non-letters in message + keyword (per starter rules)", () => {
    expect(encode("meet me by the tree!!", "sco-nes")).toBe("egsgqwtahuiljgs");
  });

  it("returns empty string if message becomes empty after normalization", () => {
    expect(encode("!!!", "scones")).toBe("");
  });

  it("throws if keyword has no letters after normalization", () => {
    expect(() => encode("meetmebythetree", "!!!")).toThrow();
    expect(() => decode("egsgqwtahuiljgs", "!!!")).toThrow();
  });
});

File: java/app/src/test/java/AlphabetCipherTest.java
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

public class AlphabetCipherTest {

    @Test
    void encodesSampleFromPrompt() {
        assertEquals("egsgqwtahuiljgs",
                AlphabetCipher.encode("meetmebythetree", "scones"));
    }

    @Test
    void decodesSampleFromPrompt() {
        assertEquals("meetmebythetree",
                AlphabetCipher.decode("egsgqwtahuiljgs", "scones"));
    }

    @Test
    void roundTripsSimpleInputs() {
        String msg = "helloworld";
        String key = "abc";
        String c = AlphabetCipher.encode(msg, key);
        assertEquals(msg, AlphabetCipher.decode(c, key));
    }

    @Test
    void normalizesCase() {
        assertEquals("egsgqwtahuiljgs",
                AlphabetCipher.encode("MeetMeByTheTree", "SCONES"));
    }

    @Test
    void ignoresNonLettersInMessageAndKeyword() {
        assertEquals("egsgqwtahuiljgs",
                AlphabetCipher.encode("meet me by the tree!!", "sco-nes"));
    }

    @Test
    void returnsEmptyIfMessageBecomesEmptyAfterNormalization() {
        assertEquals("", AlphabetCipher.encode("!!!", "scones"));
    }

    @Test
    void throwsIfKeywordHasNoLettersAfterNormalization() {
        assertThrows(IllegalArgumentException.class, () ->
                AlphabetCipher.encode("meetmebythetree", "!!!"));
        assertThrows(IllegalArgumentException.class, () ->
                AlphabetCipher.decode("egsgqwtahuiljgs", "!!!"));
    }
}
